File: src/lib/prompts.ts
// Structured response interface for AI scoring
export interface AIScoreResponse {
    analysis: {
        description: string;        // Brief brand description
        keyProducts: string[];      // Main products/services
        strengths: string[];        // Brand strengths
        weaknesses: string[];       // Brand weaknesses
    };
    scores: {
        recommendation: number;     // 0-40: Would AI recommend this brand?
        sentiment: number;          // 0-30: Overall sentiment score
        prominence: number;         // 0-20: Brand visibility/recognition
        accuracy: number;           // 0-10: Confidence in data accuracy
    };
    socialBuzz?: {
        trending: boolean;          // Is the brand trending on social media?
        sentiment: "positive" | "mixed" | "negative"; // Social media mood
        platforms: string[];        // Where it's being discussed
        summary: string;            // Brief summary of recent chatter
    };
    overallSentiment: "positive" | "neutral" | "negative";
    totalScore: number;             // 0-100
    tips: string[];                 // Improvement suggestions
}

export function generateBrandAnalysisPrompt(brand: string, category: string): string {
    const categoryContext = category && category !== "general"
        ? `in the ${category} industry/category`
        : "in their industry";

    return `Provide a comprehensive analysis of the following brand.

Brand: ${brand}
Context: ${categoryContext}

Please provide a detailed report covering:
1. What the brand is known for
2. Key products or services
3. Market reputation and sentiment
4. Notable strengths and weaknesses

Keep your response professional, balanced, and informative. If you lack specific information about this brand, state that clearly.`;
}

export function generateRecommendationPrompt(brand: string, category: string): string {
    const categoryContext = category && category !== "general"
        ? category
        : "brand in this category";

    return `A user asks: "What is the best ${categoryContext}?"

Provide a helpful recommendation response. Discuss leading options and mention ${brand} if it is a relevant and competitive choice in this space. Be balanced, objective, and informative.`;
}

// Unified scoring guidelines to ensure consistent AI scoring across all endpoints
const SCORING_GUIDELINES = `STRICT Scoring guidelines — do NOT inflate scores. Be honest and critical:
- recommendation (0-40): 35-40=top 3 in category globally, 25-34=strong recommend, 15-24=average/neutral, 5-14=concerns exist, 0-4=actively avoid
- sentiment (0-30): 26-30=universally loved, 18-25=mostly positive, 10-17=mixed reviews, 5-9=mostly negative, 0-4=PR crisis / scandals
- prominence (0-20): 18-20=global household name, 13-17=well-known nationally, 8-12=known in industry, 3-7=niche/emerging, 0-2=unknown
- accuracy (0-10): 9-10=extensive verified data, 5-8=moderate data, 2-4=limited info, 0-1=almost no data

IMPORTANT scoring rules:
- Most brands should score 40-70. Only truly exceptional global brands score 80+.
- A score of 90+ should be EXTREMELY rare (Apple, Google, Samsung tier).
- If you are unsure about a brand, score accuracy LOW (0-3) and be conservative on other metrics.
- NEVER give a perfect 100. No brand is flawless.`;

// Social media research instruction for prompts
const SOCIAL_MEDIA_RESEARCH = `SOCIAL MEDIA & FORUM RESEARCH:
Before scoring, research and consider the brand's recent presence and sentiment on:
- Reddit (subreddit mentions, reviews, complaints)
- X / Twitter (trending topics, brand mentions, customer complaints)
- Threads (recent discussions)
- TikTok (brand mentions, viral content, reviews)
- Quora (questions and answers about the brand)
- Facebook (page engagement, community sentiment)
- Instagram (brand presence, influencer mentions)
- News articles and press coverage

Factor in:
- Recent controversies, scandals, or PR issues (last 24 hours to last week)
- Customer complaints trending on social media
- Viral positive or negative content
- Recent product launches, recalls, or service outages
- Community sentiment shifts

This social research MUST influence your scores — especially sentiment and recommendation.`;

// NEW: Structured prompt that returns JSON with scores
export function generateStructuredBrandPrompt(brand: string, category: string): string {
    const categoryContext = category && category !== "general"
        ? category
        : "general";

    return `You are an expert brand intelligence analyst. Analyze the brand "${brand}" in the ${categoryContext} industry.

${SOCIAL_MEDIA_RESEARCH}

${SCORING_GUIDELINES}

You MUST respond ONLY with valid JSON matching this exact structure (no markdown, no explanation):
{
  "analysis": {
    "description": "Brief 1-2 sentence description of the brand",
    "keyProducts": ["product1", "product2", "product3"],
    "strengths": ["strength1", "strength2"],
    "weaknesses": ["weakness1", "weakness2"]
  },
  "scores": {
    "recommendation": <number 0-40>,
    "sentiment": <number 0-30>,
    "prominence": <number 0-20>,
    "accuracy": <number 0-10>
  },
  "socialBuzz": {
    "trending": <true if brand is trending on any platform, false otherwise>,
    "sentiment": "positive" | "mixed" | "negative",
    "platforms": ["list of platforms where brand is being discussed"],
    "summary": "1-2 sentence summary of what people are saying on social media right now"
  },
  "overallSentiment": "positive" | "neutral" | "negative",
  "totalScore": <number 0-100, MUST equal sum of all 4 scores>,
  "tips": ["tip 1", "tip 2", "tip 3"]
}

Be brutally honest. Do NOT inflate scores. Most brands score 40-70. Provide 3 actionable tips.`;
}

// Batch prompt: score ALL brands in an industry in a single request
export interface BatchBrandScore {
  brand: string;
  score: number;
  breakdown: {
    recommendation: number;
    sentiment: number;
    prominence: number;
    accuracy: number;
  };
}

export function generateBatchIndustryPrompt(brands: string[], category: string): string {
  const brandList = brands.map((b, i) => `${i + 1}. ${b}`).join('\n');

  return `You are an expert brand intelligence analyst. Score these ${brands.length} brands in the Indian ${category} industry for AI visibility.

${SOCIAL_MEDIA_RESEARCH}

Brands:
${brandList}

Respond ONLY with valid JSON (no markdown, no explanation):
{
  "brands": [
    {
      "brand": "Brand Name",
      "score": <number 0-100>,
      "breakdown": {
        "recommendation": <number 0-40>,
        "sentiment": <number 0-30>,
        "prominence": <number 0-20>,
        "accuracy": <number 0-10>
      }
    }
  ]
}

${SCORING_GUIDELINES}
- score MUST equal sum of all breakdown values (max 100)

Score ALL ${brands.length} brands. Be brutally honest — most should score 40-70.`;
}

export function parseBatchIndustryResponse(text: string): BatchBrandScore[] {
  try {
    let jsonStr = text.trim();
    if (jsonStr.startsWith("```json")) jsonStr = jsonStr.slice(7);
    else if (jsonStr.startsWith("```")) jsonStr = jsonStr.slice(3);
    if (jsonStr.endsWith("```")) jsonStr = jsonStr.slice(0, -3);
    jsonStr = jsonStr.trim();

    const parsed = JSON.parse(jsonStr);
    const brands = parsed.brands || parsed;

    if (!Array.isArray(brands)) return [];

    return brands.map((b: any) => ({
      brand: String(b.brand || b.name || ''),
      score: Math.min(100, Math.max(0, Number(b.score || b.totalScore) || 0)),
      breakdown: {
        recommendation: Math.min(40, Math.max(0, Number(b.breakdown?.recommendation || b.scores?.recommendation) || 0)),
        sentiment: Math.min(30, Math.max(0, Number(b.breakdown?.sentiment || b.scores?.sentiment) || 0)),
        prominence: Math.min(20, Math.max(0, Number(b.breakdown?.prominence || b.scores?.prominence) || 0)),
        accuracy: Math.min(10, Math.max(0, Number(b.breakdown?.accuracy || b.scores?.accuracy) || 0)),
      }
    })).filter((b: BatchBrandScore) => b.brand.length > 0);
  } catch (error) {
    console.warn('Failed to parse batch industry response:', error);
    return [];
  }
}

// Parse and validate AI response JSON
export function parseAIScoreResponse(text: string): AIScoreResponse | null {
    try {
        // Try to extract JSON from the response (handle markdown code blocks)
        let jsonStr = text.trim();

        // Remove markdown code blocks if present
        if (jsonStr.startsWith("```json")) {
            jsonStr = jsonStr.slice(7);
        } else if (jsonStr.startsWith("```")) {
            jsonStr = jsonStr.slice(3);
        }
        if (jsonStr.endsWith("```")) {
            jsonStr = jsonStr.slice(0, -3);
        }
        jsonStr = jsonStr.trim();

        const parsed = JSON.parse(jsonStr);

        // Validate required fields exist
        if (!parsed.analysis || !parsed.scores || !parsed.tips) {
            console.warn("Missing required fields in AI response");
            return null;
        }

        // Handle optional socialBuzz
        let socialBuzz = undefined;
        if (parsed.socialBuzz) {
            socialBuzz = {
                trending: Boolean(parsed.socialBuzz.trending),
                sentiment: ["positive", "mixed", "negative"].includes(parsed.socialBuzz.sentiment)
                    ? parsed.socialBuzz.sentiment
                    : "mixed",
                platforms: Array.isArray(parsed.socialBuzz.platforms) ? parsed.socialBuzz.platforms.map(String) : [],
                summary: String(parsed.socialBuzz.summary || "")
            };
        }

        // Validate and clamp score ranges
        const scores = {
            recommendation: Math.min(40, Math.max(0, Number(parsed.scores.recommendation) || 0)),
            sentiment: Math.min(30, Math.max(0, Number(parsed.scores.sentiment) || 0)),
            prominence: Math.min(20, Math.max(0, Number(parsed.scores.prominence) || 0)),
            accuracy: Math.min(10, Math.max(0, Number(parsed.scores.accuracy) || 0)),
        };

        const totalScore = Math.min(100, scores.recommendation + scores.sentiment + scores.prominence + scores.accuracy);

        // Validate sentiment value
        const validSentiments = ["positive", "neutral", "negative"];
        const overallSentiment = validSentiments.includes(parsed.overallSentiment)
            ? parsed.overallSentiment
            : "neutral";

        return {
            analysis: {
                description: String(parsed.analysis.description || ""),
                keyProducts: Array.isArray(parsed.analysis.keyProducts)
                    ? parsed.analysis.keyProducts.map(String)
                    : [],
                strengths: Array.isArray(parsed.analysis.strengths)
                    ? parsed.analysis.strengths.map(String)
                    : [],
                weaknesses: Array.isArray(parsed.analysis.weaknesses)
                    ? parsed.analysis.weaknesses.map(String)
                    : [],
            },
            scores,
            socialBuzz,
            overallSentiment: overallSentiment as "positive" | "neutral" | "negative",
            totalScore,
            tips: Array.isArray(parsed.tips) ? parsed.tips.map(String).slice(0, 4) : [],
        };
    } catch (error) {
        console.warn("Failed to parse AI score response:", error);
        return null;
    }
}

// HEAD-TO-HEAD COMPARISON TYPES AND FUNCTIONS

export interface BrandComparisonScores {
    recommendation: number;
    sentiment: number;
    prominence: number;
    accuracy: number;
}

export interface BrandComparisonResult {
    name: string;
    scores: BrandComparisonScores;
    totalScore: number;
    overallSentiment: "positive" | "neutral" | "negative";
}

export interface ComparisonResponse {
    brand1: BrandComparisonResult;
    brand2: BrandComparisonResult;
    comparisonSummary: string;
}

// Generate a prompt that compares both brands head-to-head in a single call
export function generateComparisonPrompt(brand1: string, brand2: string, category: string): string {
    const categoryContext = category && category !== "general" ? category : "general";

    return `You are an expert brand intelligence analyst. Compare these two brands HEAD-TO-HEAD in the ${categoryContext} industry.

${SOCIAL_MEDIA_RESEARCH}

Brand 1: ${brand1}
Brand 2: ${brand2}

IMPORTANT: You MUST score both brands using the EXACT SAME strict criteria and context. The relative scores must be consistent.
Do NOT inflate scores. Be brutally honest.

Respond ONLY with valid JSON (no markdown, no explanation):
{
  "brand1": {
    "name": "${brand1}",
    "scores": {
      "recommendation": <number 0-40>,
      "sentiment": <number 0-30>,
      "prominence": <number 0-20>,
      "accuracy": <number 0-10>
    },
    "totalScore": <sum of all scores, max 100>,
    "overallSentiment": "positive" | "neutral" | "negative"
  },
  "brand2": {
    "name": "${brand2}",
    "scores": {
      "recommendation": <number 0-40>,
      "sentiment": <number 0-30>,
      "prominence": <number 0-20>,
      "accuracy": <number 0-10>
    },
    "totalScore": <sum of all scores, max 100>,
    "overallSentiment": "positive" | "neutral" | "negative"
  },
  "comparisonSummary": "One sentence comparing the two brands"
}

${SCORING_GUIDELINES}`;
}

// Parse comparison response from AI
export function parseComparisonResponse(text: string): ComparisonResponse | null {
    try {
        let jsonStr = text.trim();

        // Remove markdown code blocks if present
        if (jsonStr.startsWith("```json")) {
            jsonStr = jsonStr.slice(7);
        } else if (jsonStr.startsWith("```")) {
            jsonStr = jsonStr.slice(3);
        }
        if (jsonStr.endsWith("```")) {
            jsonStr = jsonStr.slice(0, -3);
        }
        jsonStr = jsonStr.trim();

        const parsed = JSON.parse(jsonStr);

        if (!parsed.brand1 || !parsed.brand2) {
            console.warn("Missing brand data in comparison response");
            return null;
        }

        const parseBrandResult = (brand: unknown): BrandComparisonResult | null => {
            if (!brand || typeof brand !== 'object') return null;
            const b = brand as Record<string, unknown>;
            const scores = b.scores as Record<string, unknown> | undefined;
            if (!scores) return null;

            const parsedScores: BrandComparisonScores = {
                recommendation: Math.min(40, Math.max(0, Number(scores.recommendation) || 0)),
                sentiment: Math.min(30, Math.max(0, Number(scores.sentiment) || 0)),
                prominence: Math.min(20, Math.max(0, Number(scores.prominence) || 0)),
                accuracy: Math.min(10, Math.max(0, Number(scores.accuracy) || 0)),
            };

            const totalScore = Math.min(100,
                parsedScores.recommendation + parsedScores.sentiment +
                parsedScores.prominence + parsedScores.accuracy
            );

            const validSentiments = ["positive", "neutral", "negative"];
            const sentiment = validSentiments.includes(String(b.overallSentiment))
                ? String(b.overallSentiment) as "positive" | "neutral" | "negative"
                : "neutral";

            return {
                name: String(b.name || ""),
                scores: parsedScores,
                totalScore,
                overallSentiment: sentiment,
            };
        };

        const brand1Result = parseBrandResult(parsed.brand1);
        const brand2Result = parseBrandResult(parsed.brand2);

        if (!brand1Result || !brand2Result) {
            console.warn("Failed to parse brand results");
            return null;
        }

        return {
            brand1: brand1Result,
            brand2: brand2Result,
            comparisonSummary: String(parsed.comparisonSummary || ""),
        };
    } catch (error) {
        console.warn("Failed to parse comparison response:", error);
        return null;
    }
}

File: pipelines/prompts.py
"""
rAsh Score v2.0 — Prompt Templates & Response Parsing

Scoring approach (v2):
  AI rates each dimension 0-100, we apply weights:
    recommendation: raw/100 × 40 → 0-40
    sentiment:      raw/100 × 30 → 0-30
    prominence:     raw/100 × 20 → 0-20
    accuracy:       raw/100 × 10 → 0-10
  Total = sum of weighted scores → 0-100
"""

from __future__ import annotations
import json
import re
from config import SCORE_BOUNDS, MAX_TOTAL_SCORE
from scoring import apply_weightage


# ─── Scoring Guidelines (all dimensions 0-100) ──────────────────────────────

SCORING_GUIDELINES = """STRICT Scoring guidelines — ALL scores are 0 to 100. Be honest and critical:

- recommendation (0-100): How likely is this brand to be recommended by AI assistants?
    90-100=undisputed category leader globally, 70-89=strong recommend, 40-69=average/neutral, 15-39=concerns exist, 0-14=actively avoid

- sentiment (0-100): Overall public and social media sentiment towards this brand?
    90-100=universally loved, 60-89=mostly positive, 30-59=mixed reviews, 10-29=mostly negative, 0-9=PR crisis / scandals

- prominence (0-100): How well-known and visible is this brand?
    90-100=global household name, 65-89=well-known nationally, 35-64=known in industry, 10-34=niche/emerging, 0-9=unknown

- accuracy (0-100): How much verified data do you have about this brand?
    90-100=extensive verified data, 50-89=moderate data, 20-49=limited info, 0-19=almost no data

IMPORTANT scoring rules:
- Most brands should score 40-70 on each dimension. Only truly exceptional global brands score 80+.
- A score of 90+ on any dimension should be EXTREMELY rare.
- If you are unsure about a brand, score accuracy LOW (0-30) and be conservative on other metrics.
- NEVER give a perfect 100 on any dimension. No brand is flawless."""

SOCIAL_MEDIA_RESEARCH = """SOCIAL MEDIA & FORUM RESEARCH:
Before scoring, research and consider the brand's recent presence and sentiment on:
- Reddit (subreddit mentions, reviews, complaints)
- X / Twitter (trending topics, brand mentions, customer complaints)
- Threads (recent discussions)
- TikTok (brand mentions, viral content, reviews)
- Quora (questions and answers about the brand)
- Facebook (page engagement, community sentiment)
- Instagram (brand presence, influencer mentions)
- News articles and press coverage

Factor in:
- Recent controversies, scandals, or PR issues (last 24 hours to last week)
- Customer complaints trending on social media
- Viral positive or negative content
- Recent product launches, recalls, or service outages
- Community sentiment shifts

This social research MUST influence your scores — especially sentiment and recommendation."""


# ─── Batch Industry Prompt ───────────────────────────────────────────────────

def generate_batch_prompt(brands: list[str], category: str) -> str:
    """Generate a single prompt that scores ALL brands in an industry.
    Each dimension is scored 0-100, then we apply weightage post-response."""
    brand_list = "\n".join(f"{i+1}. {b}" for i, b in enumerate(brands))

    return f"""You are an expert brand intelligence analyst. Score these {len(brands)} brands in the Indian {category} industry for AI visibility.

{SOCIAL_MEDIA_RESEARCH}

Brands:
{brand_list}

Score EACH dimension on a scale of 0 to 100.

Respond ONLY with valid JSON (no markdown, no explanation):
{{
  "brands": [
    {{
      "brand": "Brand Name",
      "breakdown": {{
        "recommendation": <number 0-100>,
        "sentiment": <number 0-100>,
        "prominence": <number 0-100>,
        "accuracy": <number 0-100>
      }}
    }}
  ]
}}

{SCORING_GUIDELINES}

Score ALL {len(brands)} brands. Be brutally honest — most dimensions should be 40-70."""


# ─── Insight Prompts ─────────────────────────────────────────────────────────

def build_first_time_insight_prompt(
    industry_name: str,
    day1: dict,
    day2: dict,
) -> str:
    """Build insight prompt when no prior insight exists (uses 2 snapshots)."""

    def fmt(snap: dict) -> str:
        brands = snap.get("brands", [])[:10]
        return "\n".join(f"  {i+1}. {b['brand']}: {b['score']}/100" for i, b in enumerate(brands))

    return f"""You are an AI brand intelligence analyst covering India's top brands.

Analyze the AI visibility data below for the {industry_name} industry across two days and generate exactly 4–5 bullet-point insights.

Day 1 ({day1['run_date']}) – Industry avg: {day1['avg_score']}
{fmt(day1)}

Day 2 ({day2['run_date']}) – Industry avg: {day2['avg_score']}
{fmt(day2)}

Rules:
- Write exactly 4–5 bullet points
- Each bullet MUST start with a relevant emoji (📈 📉 🏆 🔄 ⚠️ 🚀 💡 🎯 etc.)
- Be specific: mention brand names and actual scores/ranks
- Note who is leading, who moved up/down, and any notable trends
- Keep each bullet to 1–2 concise sentences
- Do NOT include headers, markdown bold/italic, or any text outside the bullets
- Output ONLY the bullet list, nothing else"""


def build_update_insight_prompt(
    industry_name: str,
    prev_insight: str,
    prev_date: str,
    today: dict,
) -> str:
    """Build insight prompt with yesterday's context for continuity."""
    top_brands = today.get("brands", [])[:10]
    brands_text = "\n".join(
        f"  {i+1}. {b['brand']}: {b['score']}/100"
        + (f" (score {b['score_change']:+d})" if b.get("score_change") else "")
        + (f", rank {'▲' if b.get('rank_change', 0) > 0 else '▼'}{abs(b['rank_change'])}" if b.get("rank_change") else "")
        for i, b in enumerate(top_brands)
    )

    return f"""You are an AI brand intelligence analyst covering India's top brands.

Here is yesterday's insight ({prev_date}) for the {industry_name} industry:
{prev_insight}

Today's new data ({today['run_date']}) – Industry avg: {today['avg_score']}:
{brands_text}

Generate exactly 4–5 updated bullet-point insights reflecting what changed today compared to yesterday.

Rules:
- Write exactly 4–5 bullet points
- Each bullet MUST start with a relevant emoji (📈 📉 🏆 🔄 ⚠️ 🚀 💡 🎯 etc.)
- Be specific: mention brand names, scores, and rank movements
- Focus on changes: who moved up, who dropped, new leaders, surprising shifts
- Keep each bullet to 1–2 concise sentences
- Do NOT include headers, markdown bold/italic, or any text outside the bullets
- Output ONLY the bullet list, nothing else"""


# ─── Response Parsing ────────────────────────────────────────────────────────

def _clamp(value: float, low: int, high: int) -> int:
    """Clamp a value to integer within [low, high]."""
    return max(low, min(high, int(round(value))))


def _extract_json(text: str) -> str:
    """Strip markdown code fences from AI response text."""
    s = text.strip()
    if s.startswith("```json"):
        s = s[7:]
    elif s.startswith("```"):
        s = s[3:]
    if s.endswith("```"):
        s = s[:-3]
    return s.strip()


def parse_batch_response(text: str) -> list[dict]:
    """
    Parse a batch industry scoring response from an AI model.
    AI gives 0-100 per dimension → we apply weightage → final breakdown.
    Returns list of dicts: [{brand, score, breakdown: {recommendation, sentiment, prominence, accuracy}}]
    """
    try:
        json_str = _extract_json(text)
        parsed = json.loads(json_str)

        brands_list = parsed.get("brands", parsed)
        if not isinstance(brands_list, list):
            return []

        results = []
        for b in brands_list:
            brand_name = str(b.get("brand") or b.get("name") or "").strip()
            if not brand_name:
                continue

            breakdown_raw = b.get("breakdown") or b.get("scores") or {}

            # AI gives 0-100 for each dimension → apply weightage
            raw_scores = {
                "recommendation": float(breakdown_raw.get("recommendation", 0)),
                "sentiment": float(breakdown_raw.get("sentiment", 0)),
                "prominence": float(breakdown_raw.get("prominence", 0)),
                "accuracy": float(breakdown_raw.get("accuracy", 0)),
            }

            # Apply weightage: raw (0-100) → weighted (0-40, 0-30, 0-20, 0-10)
            weighted = apply_weightage(raw_scores)

            total = min(MAX_TOTAL_SCORE, sum(weighted.values()))

            results.append({
                "brand": brand_name,
                "score": total,
                "breakdown": weighted,
                "raw_scores": raw_scores,  # keep raw for debugging
            })

        return results

    except (json.JSONDecodeError, ValueError, TypeError) as e:
        print(f"  ⚠ Failed to parse batch response: {e}")
        return []
